exporter/uptraceexporter/translator.go

```diff
@@ -0,0 +1,221 @@
+			out = append(out, label.Int64(key, value.IntVal()))
+		case pdata.AttributeValueDOUBLE:
+			out = append(out, label.Float64(key, value.DoubleVal()))
+		case pdata.AttributeValueMAP:
+			if value, ok := mapLabelValue(value.MapVal()); ok {
+				out = append(out, label.KeyValue{
+					Key:   label.Key(key),
+					Value: value,
+				})
+			}
+		case pdata.AttributeValueARRAY:
+			if value, ok := arrayLabelValue(value.ArrayVal()); ok {
+				out = append(out, label.KeyValue{
+					Key:   label.Key(key),
+					Value: value,
+				})
+			}
+		}
+	})
+
+	return out
+}
+
+func uptraceEvents(events pdata.SpanEventSlice) []spanexp.Event {
+	if events.Len() == 0 {
+		return nil
+	}
+
+	outEvents := make([]spanexp.Event, events.Len())
+	for i := 0; i < events.Len(); i++ {
+		event := events.At(i)
+
+		out := &outEvents[i]
+		out.Name = event.Name()
+		out.Attrs = keyValueSlice(event.Attributes())
+		out.Time = int64(event.Timestamp())
+	}
+	return outEvents
+}
+
+func uptraceLinks(links pdata.SpanLinkSlice) []spanexp.Link {
+	if links.Len() == 0 {
+		return nil
+	}
+
+	outLinks := make([]spanexp.Link, links.Len())
+	for i := 0; i < links.Len(); i++ {
+		link := links.At(i)
+
+		out := &outLinks[i]
+		out.TraceID = link.TraceID().Bytes()
+		out.SpanID = asUint64(link.SpanID().Bytes())
+		out.Attrs = keyValueSlice(link.Attributes())
+	}
+	return outLinks
+}
+
+//------------------------------------------------------------------------------
+
+func mapLabelValue(m pdata.AttributeMap) (label.Value, bool) {
+	out := make(map[string]interface{}, m.Len())
+	m.ForEach(func(key string, val pdata.AttributeValue) {
+		out[key] = attrAsInterface(val)
+	})
+	return jsonLabelValue(out)
+}
+
+func arrayLabelValue(arr pdata.AnyValueArray) (label.Value, bool) {
+	if arr.Len() == 0 {
+		return label.Value{}, false
+	}
+
+	switch arrType := arr.At(0).Type(); arrType {
+	case pdata.AttributeValueSTRING:
+		out := make([]string, 0, arr.Len())
+		for i := 0; i < arr.Len(); i++ {
+			val := arr.At(i)
+			if val.Type() != arrType {
+				return label.Value{}, false
+			}
+			out = append(out, val.StringVal())
+		}
+		return label.ArrayValue(out), true
+
+	case pdata.AttributeValueBOOL:
+		out := make([]bool, 0, arr.Len())
+		for i := 0; i < arr.Len(); i++ {
+			val := arr.At(i)
+			if val.Type() != arrType {
+				return label.Value{}, false
+			}
+			out = append(out, val.BoolVal())
+		}
+		return label.ArrayValue(out), true
+
+	case pdata.AttributeValueINT:
+		out := make([]int64, 0, arr.Len())
+		for i := 0; i < arr.Len(); i++ {
+			val := arr.At(i)
+			if val.Type() != arrType {
+				return label.Value{}, false
+			}
+			out = append(out, val.IntVal())
+		}
+		return label.ArrayValue(out), true
+
+	case pdata.AttributeValueDOUBLE:
+		out := make([]float64, 0, arr.Len())
+		for i := 0; i < arr.Len(); i++ {
+			val := arr.At(i)
+			if val.Type() != arrType {
+				return label.Value{}, false
+			}
+			out = append(out, val.DoubleVal())
+		}
+		return label.ArrayValue(out), true
+	}
```

Comment:
Same as comment above. Should you have a default case here to catch unknown attributes? How about the pdata.AttributeValueNULL attribute?

Reply:
The code here is different and does a default case. The default case is it to marshal array as `[]interface{}` - see the code below.

Reply:
That makes sense. Thank you!